features/cellular/framework/AT/AT_CellularNetwork.cpp

```diff
@@ -0,0 +1,973 @@
+
+    return _at.unlock_return_error();
+#endif
+}
+
+void AT_CellularNetwork::attach(Callback<void(nsapi_event_t, intptr_t)> status_cb)
+{
+    _connection_status_cb = status_cb;
+}
+
+nsapi_connection_status_t AT_CellularNetwork::get_connection_status() const
+{
+    return _connect_status;
+}
+
+nsapi_error_t AT_CellularNetwork::set_blocking(bool blocking)
+{
+#if NSAPI_PPP_AVAILABLE
+    return nsapi_ppp_set_blocking(blocking);
+#else
+    return NSAPI_ERROR_UNSUPPORTED;
+#endif
+}
+
+
+#if NSAPI_PPP_AVAILABLE
+void AT_CellularNetwork::ppp_status_cb(nsapi_event_t event, intptr_t parameter)
+{
+    _connect_status = (nsapi_connection_status_t)parameter;
+
+    if (_connection_status_cb) {
+        _connection_status_cb(event, parameter);
+    }
+}
+#endif
+
+
+
+nsapi_error_t AT_CellularNetwork::set_context_to_be_activated()
+{
+    // try to find or create context with suitable stack
+    if (!get_context(_ip_stack_type_requested)) {
+        return NSAPI_ERROR_NO_CONNECTION;
+    }
+
+    // if user has defined user name and password we need to call CGAUTH before activating or modifying context
+    if (_pwd && _uname) {
+        _at.cmd_start("AT+CGAUTH=");
+        _at.write_int(_cid);
+        _at.write_int(_authentication_type);
+        _at.write_string(_uname);
+        _at.write_string(_pwd);
+        _at.cmd_stop();
+        _at.resp_start();
+        _at.resp_stop();
+        if (_at.get_last_error() != NSAPI_ERROR_OK) {
+            return NSAPI_ERROR_AUTH_FAILURE;
+        }
+    }
+
+    return _at.get_last_error();
+}
+
+bool AT_CellularNetwork::set_new_context(nsapi_ip_stack_t stack, int cid)
+{
+    nsapi_ip_stack_t tmp_stack = stack;
+    char pdp_type[8+1] = {0};
+
+    switch (stack) {
+        case IPV4_STACK:
+            strncpy(pdp_type, "IP", sizeof(pdp_type));
+            break;
+        case IPV6_STACK:
+            strncpy(pdp_type, "IPV6", sizeof(pdp_type));
+            break;
+        case IPV4V6_STACK:
+            strncpy(pdp_type, "IPV4V6", sizeof(pdp_type));
+            break;
+        default:
+            strncpy(pdp_type, "", sizeof(pdp_type));
+            break;
+    }
+
+    //apn: "If the value is null or omitted, then the subscription value will be requested."
+    bool success = false;
+    _at.cmd_start("AT+CGDCONT=");
+    _at.write_int(cid);
+    _at.write_string(pdp_type);
+    _at.write_string(_apn);
+    _at.cmd_stop();
+    _at.resp_start();
+    _at.resp_stop();
+    success = (_at.get_last_error() == NSAPI_ERROR_OK);
+
+    // Fall back to ipv4
+    if (!success && tmp_stack == IPV4V6_STACK) {
+        tmp_stack = IPV4_STACK;
+        _at.cmd_start("AT+FCLASS=0;+CGDCONT=");
+        _at.write_int(cid);
+        _at.write_string("IP");
+        _at.write_string(_apn);
+        _at.cmd_stop();
+        _at.resp_start();
+        _at.resp_stop();
+        success = (_at.get_last_error() == NSAPI_ERROR_OK);
+    }
+
+    if (success) {
+        _ip_stack_type = tmp_stack;
+        _cid = cid;
```

Comment:
If context created successfully it should be deleted if connect(CGACT for this context) fails.
Internal ticket: IOTCELL-538

Reply:
Fixed in: https://github.com/ARMmbed/mbed-os/pull/6082/commits/55992a177d29e34c4071ba869ccc8c62b4c526be